In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import * 

In [ ]:
import numpy as np

1.Answer\
The Mathematical model for the Assignment Problem is:\
Let $C_{ij}$ be the set up cost of $i^\text{th}$ facility at $j^\text{th}$ location  and $x_{ij}$ denotes the assignment of $i^\text{th}$ facility to $j^\text{th}$ location .Then our optimization problem is :

$\textbf{Minimize:}$
\begin{align}
\sum_{i=1}^{12}\sum_{j=1}^{12} C_{ij} x_{ij}
\end{align}

$\textbf{Subject to:}$
\begin{align}
\sum_{j} x_{ij} &= 1 &&\forall\text{  i = 1, . . . , 12}\\
\sum_{i} x_{ij} &= 1 &&\forall\text{  j = 1 , . . . , 12  }
\end{align}

$\text{where decision variables}\text{  }\  x_{ij}\text{ }\epsilon\text{ }\text{{0,1}} $

In [ ]:
cost_matrix = np.loadtxt('lab5_ex2.txt', delimiter=',')

In [ ]:
print(cost_matrix)

[[19. 12. 18. 19. 22. 21. 17. 20. 16. 15. 21. 24.]
 [22. 22. 19. 21. 22. 24. 18. 17. 21. 19. 22. 23.]
 [18. 23. 20. 20. 21. 22. 19. 18. 20. 23. 19. 19.]
 [18. 21. 20. 18. 17. 19. 24. 16. 18. 16. 20. 24.]
 [23. 17. 16. 19. 24. 21. 23. 21. 20. 21. 22. 21.]
 [23. 20. 17. 16. 20. 23. 22. 25. 24. 19. 17. 20.]
 [22. 18. 17. 15. 22. 24. 23. 20. 22. 19. 23. 20.]
 [24. 22. 21. 23. 18. 17. 16. 19. 24. 21. 20. 23.]
 [21. 20. 17. 18. 16. 24. 19. 17. 18. 20. 21. 23.]
 [19. 22. 21. 24. 20. 23. 19. 18. 23. 24. 25. 20.]
 [20. 24. 22. 20. 23. 19. 18. 16. 22. 24. 21. 24.]
 [22. 23. 24. 20. 21. 20. 20. 19. 17. 19. 20. 22.]]


In [ ]:
A=np.array(cost_matrix)

In [ ]:
model = ConcreteModel()

In [ ]:
M=12 #denoting number of factories
N=12 #denoting number of locations


In [ ]:
col_indices = np.arange(N)
row_indices = np.arange(M)
print(col_indices)
print(row_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11]
[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [ ]:
model.x = Var(row_indices, col_indices, domain= Binary)

In [ ]:
model.objective = Objective(expr=sum(A[i][j]*model.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model.constraints.add(expr=sum(model.x[i,j] for j in col_indices) == 1)

for j in col_indices:
  model.constraints.add(expr=sum(model.x[i,j] for i in row_indices) == 1)

In [ ]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# solution display
print('\nObjective = ', model.objective())

print('\nDecision Variables with non zero values')
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value != 0 :
      print(model.x[i,j],' = ',model.x[i,j].value)
    else:
      pass


Objective =  198.0

Decision Variables with non zero values
x[0,1]  =  1.0
x[1,6]  =  1.0
x[2,11]  =  1.0
x[3,9]  =  1.0
x[4,2]  =  1.0
x[5,10]  =  1.0
x[6,3]  =  1.0
x[7,5]  =  1.0
x[8,4]  =  1.0
x[9,0]  =  1.0
x[10,7]  =  1.0
x[11,8]  =  1.0


8.ANSWER:
Objective =  198.0

Facility 1 must be opened at Location 2 ;\
Facility 2 must be opened at Location 7 ;\
Facility 3 must be opened at Location 12 ;\
Facility 4 must be opened at Location 10 ;\
Facility 5 must be opened at Location 3 ;\
Facility 6 must be opened at Location 11 ;\
Facility 7 must be opened at Location 4 ;\
Facility 8 must be opened at Location 6 ;\
Facility 9 must be opened at Location 5 ;\
Facility 10 must be opened at Location 1 ;\
Facility 11 must be opened at Location 8 ;\
Facility 12 must be opened at Location 9 ;

9.Now changing the integer variables in the model to continuous variables

In [ ]:
new_model = ConcreteModel()

In [ ]:
lower_bound = 0
upper_bound = 1

In [ ]:
new_model.x = Var(row_indices, col_indices, domain= Reals)

In [ ]:
for i in row_indices: 
  for j in col_indices:
    new_model.x[i,j].setlb(lower_bound)
    new_model.x[i,j].setub(upper_bound)

In [ ]:
new_model.objective = Objective(expr=sum(A[i][j]*new_model.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [ ]:
new_model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  new_model.constraints.add(expr=sum(new_model.x[i,j] for j in col_indices) == 1)

for j in col_indices:
  new_model.constraints.add(expr=sum(new_model.x[i,j] for i in row_indices) == 1)

In [ ]:
new_model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
new_result = opt_cbc.solve(new_model)
print(new_result)

print('Solver status:', new_result.solver.status)
print('Solver termination condition:',new_result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 198.0
  Upper bound: 198.0
  Number of objectives: 1
  Number of constraints: 25
  Number of variables: 145
  Number of nonzeros: 144
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 26
  Error rc: 0
  Time: 0.017049074172973633
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal


In [ ]:
# solution display
print('\nObjective = ', new_model.objective())

print('\nDecision Variables with non zero values')
for i in row_indices:
  for j in col_indices:
    if new_model.x[i,j].value != 0 :
      print(new_model.x[i,j],' = ',new_model.x[i,j].value)
    else:
      pass


Objective =  198.0

Decision Variables with non zero values
x[0,1]  =  1.0
x[1,6]  =  1.0
x[2,11]  =  1.0
x[3,9]  =  1.0
x[4,2]  =  1.0
x[5,10]  =  1.0
x[6,3]  =  1.0
x[7,5]  =  1.0
x[8,4]  =  1.0
x[9,0]  =  1.0
x[10,7]  =  1.0
x[11,8]  =  1.0


9.$\textbf{ANSWER:}$\
Objective =  198.0 \

optimal solution is:\
$x_{12} = x_{27} = x_{3,12} = x_{4,10} = x_{53} = x_{6,11} = x_{74} = x_{86} = x_{95} = x_{10,1} = x_{11,8} = x_{12,9} = 1$


10.$\textbf{ANSWER:}$\
The optimal costs for both problems are same i.e. 198 .\
In both cases the values of the variables are
integer-valued

12.$\textbf{ANSWER:}$\
If facility 1 cannot be assigned to location 4, facility
11 cannot be assigned to location 3 and facility 5 cannot be assigned to location 9.Then in that case we will assume the decision variables $x_{14}$ , $x_{11,3}$ , $x_{59}$  $\epsilon\text{ }\text{{0}}$  in our optimization model.\
$Alternatively$ we can increase the cost (in the text file) corresponding to these variables to a level (here 35) so that these variables are automatically neglected while obtaining the optimum cost.\
Here i choose to procede with the later approach.

In [ ]:
new_cost_matrix = np.loadtxt('lab5_ex2_1.txt', delimiter=',')

In [ ]:
print(new_cost_matrix)

[[19. 12. 18. 35. 22. 21. 17. 20. 16. 15. 21. 24.]
 [22. 22. 19. 21. 22. 24. 18. 17. 21. 19. 22. 23.]
 [18. 23. 20. 20. 21. 22. 19. 18. 20. 23. 19. 19.]
 [18. 21. 20. 18. 17. 19. 24. 16. 18. 16. 20. 24.]
 [23. 17. 16. 19. 24. 21. 23. 21. 35. 21. 22. 21.]
 [23. 20. 17. 16. 20. 23. 22. 25. 24. 19. 17. 20.]
 [22. 18. 17. 15. 22. 24. 23. 20. 22. 19. 23. 20.]
 [24. 22. 21. 23. 18. 17. 16. 19. 24. 21. 20. 23.]
 [21. 20. 17. 18. 16. 24. 19. 17. 18. 20. 21. 23.]
 [19. 22. 21. 24. 20. 23. 19. 18. 23. 24. 25. 20.]
 [20. 24. 35. 20. 23. 19. 18. 16. 22. 24. 21. 24.]
 [22. 23. 24. 20. 21. 20. 20. 19. 17. 19. 20. 22.]]


In [ ]:
E=np.array(new_cost_matrix)

In [ ]:
model_1 = ConcreteModel()

In [ ]:
M=12 #denoting number of factories
N=12 #denoting number of locations

In [ ]:
col_indices = np.arange(N)
row_indices = np.arange(M)
print(col_indices)
print(row_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11]
[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [ ]:
model_1.x = Var(row_indices, col_indices, domain= Binary)

In [ ]:
model_1.objective = Objective(expr=sum(E[i][j]*model_1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [ ]:
model_1.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model_1.constraints.add(expr=sum(model_1.x[i,j] for j in col_indices) == 1)

for j in col_indices:
  model_1.constraints.add(expr=sum(model_1.x[i,j] for i in row_indices) == 1)

In [ ]:
model_1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result_2 = opt_cbc.solve(model)
print('Solver status:', result_2.solver.status)
print('Solver termination condition:',result_2.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


12.ANSWER:\
Objective = 198.0

Decision Variables with non zero values\
x[0,1]  =  1.0 \
x[1,6]  =  1.0 \
x[2,11]  =  1.0 \
x[3,9]  =  1.0 \
x[4,2]  =  1.0 \
x[5,10]  =  1.0 \
x[6,3]  =  1.0 \
x[7,5]  =  1.0 \
x[8,4]  =  1.0 \
x[9,0]  =  1.0 \
x[10,7]  =  1.0 \
x[11,8]  =  1.0 \
